File: disk_sense/undo_manager.py
"""SQLite 操作日志与批量原子性。

每次文件操作先落日志（ACTIVE）再执行，随后更新 DONE/FAILED——
先日志后执行保证进程中途崩溃也能追溯。批量操作共享一个 op_uuid，
回滚时整批逐条执行、失败跳过并汇总。

线程模型：FastAPI 的 to_thread 会从不同线程调用，连接以
check_same_thread=False 创建并全程持锁串行化。
"""

from __future__ import annotations

import gzip
import json
import logging
import sqlite3
import threading
from datetime import datetime, timedelta
from pathlib import Path
from typing import Any, Optional

logger = logging.getLogger(__name__)

_SCHEMA = """
CREATE TABLE IF NOT EXISTS operation_log (
    id INTEGER PRIMARY KEY AUTOINCREMENT,
    op_uuid TEXT NOT NULL,
    session_id TEXT,
    op_type TEXT NOT NULL,
    source_path TEXT NOT NULL,
    dest_path TEXT,
    file_size INTEGER,
    file_mtime REAL,
    recycle_bin_name TEXT,
    recycle_info_name TEXT,
    recycle_path TEXT,
    status TEXT DEFAULT 'ACTIVE',
    created_at TEXT DEFAULT (datetime('now', 'localtime')),
    undone_at TEXT,
    error_msg TEXT
);
CREATE INDEX IF NOT EXISTS idx_op_created ON operation_log (created_at DESC);
CREATE INDEX IF NOT EXISTS idx_op_uuid ON operation_log (op_uuid);
"""

# 可回滚状态：执行成功（DONE）或尚未执行（ACTIVE）的记录
_UNDOABLE = ("ACTIVE", "DONE")


class UndoManager:
    """操作日志持久化与查询。"""

    def __init__(self, db_path: Path | str, retention_days: int = 30):
        self.db_path = Path(db_path)
        self.retention_days = retention_days
        self.db_path.parent.mkdir(parents=True, exist_ok=True)
        self._lock = threading.Lock()
        self._conn = sqlite3.connect(str(self.db_path), check_same_thread=False)
        self._conn.row_factory = sqlite3.Row
        with self._lock, self._conn:
            self._conn.executescript(_SCHEMA)

    def close(self) -> None:
        with self._lock:
            self._conn.close()

    # ------------------------------------------------------------------
    def log_batch(
        self,
        op_uuid: str,
        op_type: str,
        entries: list[dict[str, Any]],
        session_id: Optional[str] = None,
    ) -> list[int]:
        """插入一批操作日志（status=ACTIVE），返回自增 id 列表。"""
        ids: list[int] = []
        with self._lock, self._conn:
            for e in entries:
                cur = self._conn.execute(
                    """INSERT INTO operation_log
                       (op_uuid, session_id, op_type, source_path, dest_path, file_size, file_mtime)
                       VALUES (?, ?, ?, ?, ?, ?, ?)""",
                    (
                        op_uuid,
                        session_id,
                        op_type,
                        e["source_path"],
                        e.get("dest_path"),
                        e.get("file_size"),
                        e.get("file_mtime"),
                    ),
                )
                ids.append(cur.lastrowid)
        return ids

    def update_entry(self, op_id: int, **fields: Any) -> None:
        """按列名更新记录（status/recycle_*/error_msg/undone_at...）。"""
        if not fields:
            return
        cols = ", ".join(f"{k} = ?" for k in fields)
        with self._lock, self._conn:
            self._conn.execute(
                f"UPDATE operation_log SET {cols} WHERE id = ?", (*fields.values(), op_id)
            )

    # ------------------------------------------------------------------
    def get_entry(self, op_id: int) -> Optional[sqlite3.Row]:
        with self._lock:
            return self._conn.execute(
                "SELECT * FROM operation_log WHERE id = ?", (op_id,)
            ).fetchone()

    def get_batch(self, op_uuid: str) -> list[sqlite3.Row]:
        with self._lock:
            return self._conn.execute(
                "SELECT * FROM operation_log WHERE op_uuid = ? ORDER BY id", (op_uuid,)
            ).fetchall()

    def list_ops(self, limit: int = 10) -> list[dict]:
        """最近操作（倒序扁平记录，供 Agent 审计）。"""
        with self._lock:
            rows = self._conn.execute(
                "SELECT * FROM operation_log ORDER BY id DESC LIMIT ?", (limit,)
            ).fetchall()
        return [dict(r) for r in rows]

    # ------------------------------------------------------------------
    def archive_expired(self, archive_dir: Path) -> int:
        """超期日志转存 JSON.gz 并删除（服务空闲时调用）。"""
        cutoff = (datetime.now() - timedelta(days=self.retention_days)).strftime(
            "%Y-%m-%d %H:%M:%S"
        )
        with self._lock:
            rows = self._conn.execute(
                "SELECT * FROM operation_log WHERE created_at < ?", (cutoff,)
            ).fetchall()
            if not rows:
                return 0
            archive_dir.mkdir(parents=True, exist_ok=True)
            out = archive_dir / f"op_log_{datetime.now():%Y%m%d_%H%M%S}.json.gz"
            payload = [dict(r) for r in rows]
            with gzip.open(out, "wt", encoding="utf-8") as f:
                json.dump(payload, f, ensure_ascii=False, indent=1)
            ids = [r["id"] for r in rows]
            self._conn.executemany(
                f"DELETE FROM operation_log WHERE id = ?", [(i,) for i in ids]
            )
            self._conn.commit()
        logger.info("归档 %d 条超期操作日志 → %s", len(ids), out)
        return len(ids)

File: src/operator/undo-manager.ts
/**
 * SQLite 操作日志与批量原子性。
 *
 * 每次文件操作先落日志（ACTIVE）再执行，随后更新 DONE/FAILED——
 * 先日志后执行保证进程中途崩溃也能追溯。批量操作共享一个 op_uuid，
 * 回滚时整批逐条执行、失败跳过并汇总。
 */

import fs from "node:fs";
import path from "node:path";
import { gzipSync } from "node:zlib";
import Database from "better-sqlite3";

const SCHEMA = `
CREATE TABLE IF NOT EXISTS operation_log (
    id INTEGER PRIMARY KEY AUTOINCREMENT,
    op_uuid TEXT NOT NULL,
    session_id TEXT,
    op_type TEXT NOT NULL,
    source_path TEXT NOT NULL,
    dest_path TEXT,
    file_size INTEGER,
    file_mtime REAL,
    recycle_bin_name TEXT,
    recycle_info_name TEXT,
    recycle_path TEXT,
    status TEXT DEFAULT 'ACTIVE',
    created_at TEXT DEFAULT (datetime('now', 'localtime')),
    undone_at TEXT,
    error_msg TEXT
);
CREATE INDEX IF NOT EXISTS idx_op_created ON operation_log (created_at DESC);
CREATE INDEX IF NOT EXISTS idx_op_uuid ON operation_log (op_uuid);
`;

export interface OpRow {
  id: number;
  op_uuid: string;
  session_id: string | null;
  op_type: string;
  source_path: string;
  dest_path: string | null;
  file_size: number | null;
  file_mtime: number | null;
  recycle_bin_name: string | null;
  recycle_info_name: string | null;
  recycle_path: string | null;
  status: string;
  created_at: string;
  undone_at: string | null;
  error_msg: string | null;
}

export interface LogEntry {
  source_path: string;
  dest_path?: string | null;
  file_size?: number | null;
  file_mtime?: number | null;
}

/** 操作日志持久化与查询。 */
export class UndoManager {
  private db: Database.Database;

  constructor(
    dbPath: string,
    public retentionDays = 30
  ) {
    fs.mkdirSync(path.dirname(dbPath), { recursive: true });
    this.db = new Database(dbPath);
    this.db.pragma("journal_mode = WAL");
    this.db.exec(SCHEMA);
  }

  close(): void {
    this.db.close();
  }

  /** 插入一批操作日志（status=ACTIVE），返回自增 id 列表。 */
  logBatch(
    opUuid: string,
    opType: string,
    entries: LogEntry[],
    sessionId?: string | null
  ): number[] {
    const stmt = this.db.prepare(
      `INSERT INTO operation_log
       (op_uuid, session_id, op_type, source_path, dest_path, file_size, file_mtime)
       VALUES (?, ?, ?, ?, ?, ?, ?)`
    );
    const ids: number[] = [];
    const tx = this.db.transaction(() => {
      for (const e of entries) {
        const r = stmt.run(
          opUuid, sessionId ?? null, opType,
          e.source_path, e.dest_path ?? null,
          e.file_size ?? null, e.file_mtime ?? null
        );
        ids.push(Number(r.lastInsertRowid));
      }
    });
    tx();
    return ids;
  }

  /** 按列名更新记录（status / recycle_xxx / error_msg / undone_at 等）。 */
  updateEntry(opId: number, fields: Record<string, unknown>): void {
    if (!fields || Object.keys(fields).length === 0) return;
    const cols = Object.keys(fields)
      .map((k) => `${k} = ?`)
      .join(", ");
    this.db.prepare(`UPDATE operation_log SET ${cols} WHERE id = ?`).run(
      ...Object.values(fields), opId
    );
  }

  getEntry(opId: number): OpRow | null {
    return (
      (this.db.prepare("SELECT * FROM operation_log WHERE id = ?").get(opId) as
        | OpRow
        | undefined) ?? null
    );
  }

  getBatch(opUuid: string): OpRow[] {
    return this.db
      .prepare("SELECT * FROM operation_log WHERE op_uuid = ? ORDER BY id")
      .all(opUuid) as OpRow[];
  }

  /** 最近操作（倒序扁平记录，供 Agent 审计）。 */
  listOps(limit = 10): OpRow[] {
    return this.db
      .prepare("SELECT * FROM operation_log ORDER BY id DESC LIMIT ?")
      .all(limit) as OpRow[];
  }

  /** 超期日志转存 .json.gz 并删除。 */
  archiveExpired(archiveDir: string): number {
    // 与 SQLite datetime('now','localtime') 同一坐标系：本地时间
    const d = new Date(Date.now() - this.retentionDays * 86400_000);
    const p = (n: number) => String(n).padStart(2, "0");
    const cutoff = `${d.getFullYear()}-${p(d.getMonth() + 1)}-${p(d.getDate())} ${p(d.getHours())}:${p(d.getMinutes())}:${p(d.getSeconds())}`;
    const rows = this.db
      .prepare("SELECT * FROM operation_log WHERE created_at < ?")
      .all(cutoff) as OpRow[];
    if (rows.length === 0) return 0;

    fs.mkdirSync(archiveDir, { recursive: true });
    const stamp = new Date()
      .toISOString()
      .replace(/[-:T]/g, "")
      .slice(0, 14);
    const out = path.join(archiveDir, `op_log_${stamp}.json.gz`);
    fs.writeFileSync(out, gzipSync(Buffer.from(JSON.stringify(rows, null, 1), "utf-8")));

    const del = this.db.prepare("DELETE FROM operation_log WHERE id = ?");
    const tx = this.db.transaction(() => {
      for (const r of rows) del.run(r.id);
    });
    tx();
    return rows.length;
  }
}
